### PR TITLE
fix for relative URI detection

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
ticket 3853

Fix for check on relative URI

### CODE REVIEW

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -5717,7 +5717,7 @@ static int DecodeAltNames(byte* input, int sz, DecodedCert* cert)
             /* Verify RFC 5280 Sec 4.2.1.6 rule:
                 "The name MUST NOT be a relative URI" */
 
-            if (XSTRNCMP((const char*)&input[idx], "://", strLen + 1) != 0) {
+            if (XSTRNSTR((const char*)&input[idx], "://", strLen + 1) == NULL) {
```
Comment (toddouska) [inline wolfcrypt/src/asn.c]: strcmp starts at the beginning for comparison, strstr will find the needle anywhere in the haystack.  Doesn't an absolute uri have :// at the beginning?

Comment (JacobBarthelmeh) [reply wolfcrypt/src/asn.c]: The problem encountered is that the URI scheme is not accounted for. For example http://example.com is an absolute URI, and the idx value for input buffer with this comparison was starting at 'h'. My first solution was searching for the first instance of a ':' character encountered but that became similar to just doing a strnstr. I can get it to be more exact, having logic to test that a ':' char is encountered before any '/' char and incrementing an index to skip over scheme's like http, ftp, ldap... then doing the comparision on the first encounter of ':'

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -5717,9 +5717,28 @@ static int DecodeAltNames(byte* input, int sz, DecodedCert* cert)
...
 
-            if (XSTRNCMP((const char*)&input[idx], "://", strLen + 1) != 0) {
-                WOLFSSL_MSG("\tAlt Name must be absolute URI");
-                return ASN_ALT_NAME_E;
+            {
+                int i;
+
+                /* skip past scheme (i.e http,ftp,...) finding first ':' char */
+                for (i = 0; i < strLen; i++) {
+                    if (input[idx + i] == ':') {
```
[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply wolfcrypt/src/asn.c]: thanks this is a good catch. Will add a sanity check on strLen + idx before reading from input buffer

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -5717,9 +5717,28 @@ static int DecodeAltNames(byte* input, int sz, DecodedCert* cert)
...
+                    if (input[idx + i] == '/') {
+                        i = strLen; /* error, found relative path since '/' was
+                                     * encountered before ':'. Returning error
+                                     * value in next if statement. */
+                    }
+                }
+
+                /* test if no ':' char was found and test that the next two
+                 * chars are // to match the pattern "://" */
+                if (i == strLen || (input[idx + i + 1] != '/' ||
```
[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply wolfcrypt/src/asn.c]: this is a valid point -- will add a fix

**Pull request merged**